example/bufferCopy/src/bufferCopy.cpp

```diff
@@ -55,6 +55,24 @@ struct PrintBufferKernel
     }
 };
 
+//#############################################################################
+//! Prints all elements of the buffer.
+struct PrintBufferKernel3D
+{
+    //-----------------------------------------------------------------------------
+    template<typename TAcc, typename TAccessor, typename TExtent>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc, TAccessor const data, TExtent const& extents) const
+        -> void
```

Comment:
Discussed on VC, writing to document.

1. Some of us were in favor of having size and accessor in same object, since we already have an object. Also like `std::span` does it.

2. There was a discussion of template parameter passing in style of `TAccessor` vs. `alpaka::Accessor<T, ... if needed>`. I think here it's one of our typical choices of flexibility vs. ease of use. `TAccessor` is super flexible, but not very understandable (what's the interface of this?), at least before we have proper concepts. Additionally, an explicit `Accessor` would have more template parameters than `T` probably.

3. I am not a fan that this interface does not show if `data` is read-only or read-write. This `const` currently means same as in `T * const`, only applies to pointer, but not to data. The current raw pointer interface has this distinction as in `T *` vs. `T const *`, and I don't think we should be less expressive than this when going for higher level abstractions. To achieve it one may change the semantics of `TAccessor const data` to mean constness of data instead of constness of pointer. Technically this is probably not that difficult to do, the question is more if we want it this way. But I strongly suggest to make it some way. Taking into account point 2, with a change of style there it could be expressed like `Accessor<T>` vs. `ConstAccessor<T>` to mimic raw pointer style distinction.

Reply:
> ... Accessor<T> vs. ConstAccessor<T> to mimic raw pointer style distinction.

IMO there should be a trait to get the information if the object is const or not, there is no need to mangle it into the name of the object, but it is possible.
I think also that an accessor requires well-chosen policies to be flexible. `Dim` and `PointerType` is not enough to cover textures, cudaArrays, ...

Reply:
> > ... Accessor vs. ConstAccessor to mimic raw pointer style distinction.
> 
> IMO there should be a trait to get the information if the object is const or not, there is no need to mangle it into the name of the object, but it is possible.
> I think also that an accessor requires well-chosen policies to be flexible. `Dim` and `PointerType` is not enough to cover textures, cudaArrays, ...

The question is if there is a need to define the template interface of `Accessor` or if the accessor is a concept that is implementation specific but should define traits to be able to derive information.

@bernhardmgruber Is the accessor in your example implementation a suggestion for a template signature or an example of an object that follows the concept "accessor"?

Reply:
> IMO there should be a trait to get the information if the object is const or not, there is no need to mangle it into the name of the object, but it is possible.

This is exactly NOT what I (and Sergei) want. If the accessor type was completely opaque, how can I see in a kernel interface, if `data` will we used read-only, or read-write? Of course, I could query this information. But I want to see it at the interface.

```c++
template<typename TAcc, typename TAccessor, typename TExtent>
ALPAKA_FN_ACC auto operator()(TAcc const& acc, TAccessor data, TExtent const& extents) const;
```

> The question is if there is a need to define the template interface of Accessor or if the accessor is a concept that is implementation specific but should define traits to be able to derive information.

Contrary to what we usually do in alpaka, I try to design an explicit `Accessor` type. We could still go for a concept later, but I am again worried, that we loose expressiveness in the kernel function interface.

Reply:
> I think also that an accessor requires well-chosen policies to be flexible. Dim and PointerType is not enough to cover textures, cudaArrays, ...

If we separate `Pointer` and `Value` types of an accessor, we can support textures. CUDA buffers will have `T*` and `T`, SYCL will have `sycl::global_ptr<T>`/`sycl::accessor<T, ...>` and `T`. CUDA textures will have `cudaTextureObject_t` and `T`, SYCL images will have `sycl::accessor<T, ...>` and `T`.

Reply:
First, I wanted to make a joke about `BlobStorage` in LLAMA, because I thought `Pointer` also models the same concept. But it turned out, the `Pointer` here is a different, because it also contains the object type of the sequence used for storage. LLAMA just uses `std::byte`, thus untyped storage.
Then I thought, `Pointer` is actually just another wrapped accessor. Because for SYCL, it actually is an accessor. And we could back an ND `alpaka::accessor` by a 1D `alpaka::accessor`.
So then I thought `TypedIndexableStorage`. But to support byte pitches, we actually need contigous storage which allows to form references into it, so now we are at `ContiguousTypedIndexableStorage` which must ensure that `operator[]` returns an l-value reference. But since we are going to hast that to a byte pointer anyway, we could just turn `Pointer` into `std::byte*` and be done with it.